### ISSUE TITLE
Engagement Time is logged incorrectly when disable screen view tracking

### ISSUE THREAD

**Issue by chalermpong** [firebase/firebase-android-sdk#3240]
### [READ] Step 1: Are you in the right place?
Yes.

### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: Android Studio Arctic Fox | 2020.3.1 Patch 3
  * Firebase Component: Analytics
  * Component version: 20.0.1 (Firebase BoM 29.0.2) 

### [REQUIRED] Step 3: Describe the problem

This issue is related to my previous issue #3039 
Firebase Analytics logs incorrect engagement time. As shown in Log below
```
2021-12-14 13:42:21.554 16014-16064/my.app.package I/FA: App measurement initialized, version: 46000
2021-12-14 13:42:21.554 16014-16064/my.app.package I/FA: To enable debug logging run: adb shell setprop log.tag.FA VERBOSE
2021-12-14 13:42:21.554 16014-16064/my.app.package I/FA: Faster debug mode event logging enabled. To disable, run:
adb shell setprop debug.firebase.analytics.app .none.
2021-12-14 13:42:21.554 16014-16064/my.app.package D/FA: Debug-level message logging enabled
2021-12-14 13:42:21.580 16014-16064/my.app.package V/FA: Connecting to remote service
2021-12-14 13:42:21.590 16014-16064/my.app.package V/FA: Connection attempt already in progress
2021-12-14 13:42:21.595 16014-16064/my.app.package V/FA: Connection attempt already in progress
2021-12-14 13:42:21.596 16014-16064/my.app.package D/FA: Connected to remote service
2021-12-14 13:42:21.596 16014-16064/my.app.package V/FA: Processing queued up service tasks: 3
2021-12-14 13:42:26.625 16014-16064/my.app.package V/FA: Inactivity, disconnecting from the service
2021-12-14 13:42:50.656 16014-16055/my.app.package V/FA: onActivityCreated
2021-12-14 13:42:50.700 16014-16064/my.app.package V/FA: Activity resumed, time: 246386362
2021-12-14 13:42:50.701 16014-16055/my.app.package V/FA: Logging screen view with name, class: main, MainActivity1
2021-12-14 13:42:50.705 16014-16064/my.app.package I/FA: Tag Manager is not found and thus will not be used
2021-12-14 13:42:50.728 16014-16064/my.app.package V/FA: Connecting to remote service
2021-12-14 13:42:50.745 16014-16064/my.app.package V/FA: Connection attempt already in progress
2021-12-14 13:42:50.799 16014-16064/my.app.package D/FA: Connected to remote service
2021-12-14 13:42:50.802 16014-16064/my.app.package V/FA: Processing queued up service tasks: 2
2021-12-14 13:42:55.810 16014-16064/my.app.package V/FA: Inactivity, disconnecting from the service
2021-12-14 13:42:56.270 16014-16064/my.app.package V/FA: Activity paused, time: 246391932
2021-12-14 13:42:58.272 16014-16064/my.app.package D/FA: Application going to the background
2021-12-14 13:42:58.283 16014-16064/my.app.package V/FA: Recording user engagement, ms: 34745
```
**Explanation for the log above**
```
I kill the app before start of the log above.
13:42:21.554 <-- I sent a data message to the app
13:42:50.700 <-- I started the app
13:42:56.270 <-- I left the app using Android home button
```
The engagement time should be around 6 seconds, but Firebase Analytics logged 34745 milliseconds. Seems like FA started logging engagement time when app received a data message.

#### Steps to reproduce:

**Prerequisite:** The app must [disable screen view tracking](https://firebase.google.com/docs/analytics/screenviews#disable_screenview_tracking), and be able to receive [data message](https://firebase.google.com/docs/cloud-messaging/concept-options#notifications_and_data_messages)
1. Start app and kill the app quickly before FA logs engagement time (explanation in note below).
2. Send a data message to the app
3. Open app again
4. Stay in the first screen for a few seconds then leave the app. FA will log engagement time = time since it received the data message instead of time since app open. 

**Note:** When disable screen view tracking, there will be ~2 seconds delay before FA logs engagement time. So it is possible to kill the app quickly before `Recording user engagement` happens.
```
2021-12-14 13:42:56.270 16014-16064/my.app.package V/FA: Activity paused, time: 246391932
2021-12-14 13:42:58.272 16014-16064/my.app.package D/FA: Application going to the background
2021-12-14 13:42:58.283 16014-16064/my.app.package V/FA: Recording user engagement, ms: 34745
```
If the app is not disable screen view tracking, the `Recording user engagement, ms: ...` will happen before `Activity paused, time: ...`

**Comment by argzdev** [comment]
Hi @chalermpong, thanks for reporting. I'll try to check this and see what I can find. In the meantime, if you can provide us with an [MCVE](https://stackoverflow.com/help/minimal-reproducible-example), that'll greatly help us with the investigation. Thanks!

**Comment by chalermpong** [comment]
@argzdev Here is the code that can reproduce this issue.
https://github.com/chalermpong/FirebaseEngagementTime

**Comment by argzdev** [comment]
Thanks for the MCVE, @chalermpong. I was able to repro the issue, I'll notify an engineer and see what we can do here.

**Comment by argzdev** [comment]
Our engineers have been notified and will look into this. This is now internally tracked on b/210863612. Thanks!

**Comment by OksanaAxonista** [comment]
Hello, I have the same issue.
Since 26 August our engagement times started to rise significantly only on Android
It happens with every screen. 
Here are [screenshot](https://i.stack.imgur.com/h8HkX.png) from GA for splash.

**Comment by pelevinma** [comment]
I have the same problem. In August 2021 I disabled google_analytics_automatic_screen_reporting_enabled and my in-app engagement time doubled. I would like to know if this is a bug or a consequence of using the google_analytics_automatic_screen_reporting_enabled flag?

**Comment by woodii** [comment]
We have the exact same issue. We implemented manual screen tracking on both our apps (android and ios) but in android app the engagement time (average and per session) skyrocketed by 700% from that release on.
![download](https://user-images.githubusercontent.com/1383203/163533024-fd99f9e4-ad56-4884-9d3b-4b40cfa01932.png)

**Comment by tompesman** [comment]
Hi, same problem here.
We've compared to our previous version and to iOS. 

![Screenshot 2022-07-26 at 10 45 41](https://user-images.githubusercontent.com/116475/180967827-f3016464-3dd6-47bd-8f08-96a9d12f6b71.png)

Can we expect this to be solved or should we enable automatic screen reporting to get our engagement stats back?

Cheers!

**Comment by chalermpong** [comment]
@tompesman I posted my workaround [here](https://stackoverflow.com/a/70488316/9527268). Let's see if it's work on your project.

<img width="846" alt="image" src="https://user-images.githubusercontent.com/4977805/181064793-ef07e037-02b0-4564-bd6d-9bdd6f730222.png">

**Comment by tompesman** [comment]
That's an interesting one, are you sure that there are no side effects?

**Comment by chalermpong** [comment]
@tompesman You are right. I'm not sure if there is any side effect.

@argzdev Could you help us confirm if there are no side effects from using this workaround?

Instead of adding this line in AndroidManifest.xml,
```xml
<meta-data android:name="google_analytics_automatic_screen_reporting_enabled" android:value="false" />
``` 
I used `setAnalyticsCollectionEnabled` in Activity's `onResume` to disable automatic screen tracking as shown in code below.
```kotlin
override fun onResume() {
    FirebaseAnalytics.getInstance(this).setAnalyticsCollectionEnabled(false)
    super.onResume()
    FirebaseAnalytics.getInstance(this).setAnalyticsCollectionEnabled(true)
}
```

**Comment by argzdev** [comment]
Hi all, thank you for the inputs! Unfortunately, I can't verify if this workaround has no side effects. 

I have escalated this to our engineering team, and it looks like our engineers are still investigating this along with other issues and tasks. Given that their hands are full at the moment, it my take some time for them to dig deeper in this issue. That being said, while I’m unable to promise any timeline for this, adding a thumbs up on this issue can help our engineers prioritize this based on the demand and severity.

**Comment by leontodd** [comment]
Also having the same issue, which makes engagement statistics useless.

@argzdev How do I thumbs up an issue on GitHub? Do you mean using the reaction emojis?

**Comment by argzdev** [comment]
Yes @leontodd, apologies if that was confusing.

**Comment by jaumebd** [comment]
Hi, @argzdev there's any update on this issue? I'm experimenting the same misbehaviour after disabling automatic screen reporting

**Comment by jianglanCao** [comment]
Are there some new updates？
This issue is still on our apps.

**Comment by mypplication** [comment]
Hello,
Same problem here...

<img width="529" alt="image" src="https://user-images.githubusercontent.com/1562126/215422885-afb01152-fd4d-4b9f-bd26-9122516f7bb6.png">

**Comment by TaylorPringleNZME** [comment]
Hi, Is there any update on this issue? @argzdev

**Comment by argzdev** [comment]
Hi @TaylorPringleNZME, thanks for following up on us. Looking into the internal bug, it seems that there's a fix on going, though I'm not sure if this has been merged on the latest version. I'll try to get some updates from our engineers. But if possible, please do try using our latest version to see if the fix has been applied. Thanks!

**Comment by TaylorPringleNZME** [comment]
> Hi @TaylorPringleNZME, thanks for following up on us. Looking into the internal bug, it seems that there's a fix on going, though I'm not sure if this has been merged on the latest version. I'll try to get some updates from our engineers. But if possible, please do try using our latest version to see if the fix has been applied. Thanks!

Hi @argzdev, is there any way we can track the fix? Thanks for the update

**Comment by argzdev** [comment]
Sorry @TaylorPringleNZME, currently the bug details are only available internally, but feel free to ping here whenever you'd like to get an update, and we'll take a look into it.

**Comment by TaylorPringleNZME** [comment]
Hi @argzdev, thank you for the update. 

Are you able to confirm if the fix was fully rolled out last week please? Also when would we expect to see a change in results?

**Comment by debaerebart** [comment]
Hi all,

I am writing down an analytics solution for a client of ours and I'm stuck on below question (and can't seem to find a clear answer anywhere): 

Does your engagement time still get logged if you disable the automatic screen view tracking?

The plan is to implement the screen_view event ourselves, however in the example code the user_engagement_msec is not added to the custom event (https://firebase.google.com/docs/analytics/screenviews). 

So, I am wondering is that intentional or is it just to keep the example simple?

I bumped into this post, and I know my question is not entirely on topic, but it thought that somebody in this audience might know the answer.

Appreciate it if somebody can give a clear answer!

**Comment by argzdev** [comment]
Yes @TaylorPringleNZME, the fix should be fully rolled out by now. You should start seeing a change in results immediately.

@debaerebart it would be best if you reach out to our [Firebase support team](https://firebase.google.com/support/troubleshooter/contact). They'll be able to clarify any questions you may have.

That said, since this issue have been resolved. I'll close this now. If anyone experiences this issue again, just let me know and we'll reopen this thread. Thanks!

**Comment by woodii** [comment]
@argzdev are you sure this will take effect immediately?
I don't see any changes on our side.
![image](https://github.com/firebase/firebase-android-sdk/assets/1383203/97a65dd3-3452-4c32-9ed9-87b2e12c4921)

**Comment by mthormann-dhi** [comment]
Same here. No change at all.

**Comment by TaylorPringleNZME** [comment]
Hi @argzdev, we have also not seen any changes in the past few days.

**Comment by lbarrettanderson** [comment]
> @argzdev are you sure this will take effect immediately? I don't see any changes on our side. ![image](https://user-images.githubusercontent.com/1383203/245511634-97a65dd3-3452-4c32-9ed9-87b2e12c4921.png)

@woodii  The fix is at 50% rollout, so by now you should see a significant improvement. Since you aren't, can you share with me your package name so I can research further?

**Comment by jmartinMone** [comment]
@argzdev @lbarrettanderson 
Hello, no changes here.
<img width="1470" alt="image" src="https://github.com/firebase/firebase-android-sdk/assets/130645591/f78a3e79-60ea-4548-95c1-c06a7dab5730">

one.manager.app

**Comment by lbarrettanderson** [comment]
Thank you that's really helpful.

Based on a quick look, I do see a slight improvement between the two experiment arms, but clearly not as much as expected. It seems the fix only addresses one cause of the issue. I'll investigate further.

**Comment by woodii** [comment]
Hello @argzdev @lbarrettanderson thanks for taking care of this issue.
Still no change in logged times.
at.wienerlinien.wienmobillab is our biggest app.

If you need any further information or implementation details, let me know.

**Comment by lbarrettanderson** [comment]
Thank you. I'm investigating. If you are able to reproduce this issue *without* killing the app process, please let me know what steps you're taking. The fix that's currently at 50% rollout seems to fix the repro steps described in this issue's OP.

**Comment by woodii** [comment]
Today I had some time to look into this again. Unfortunately I coudln't find any hint, but maybe my research helps.

If i check the analytics logs via 
```
adb shell setprop log.tag.FA VERBOSE
adb shell setprop log.tag.FA-SVC VERBOSE
```
I couldn't find any faulty "recording user engagement" entries. All measurements look fine.

But if I check my engagement reports all (not just some screens) have way higher engagement times than they should have.

**Comment by ninniuz** [comment]
We are having the same issue on different apps where automatic screen view tracking has been disabled.

I don't see any odd values for user engagement in logcat but reports are completely out of scale now.

Do you have any updates?

**Comment by Eluch** [comment]
Same issue here. Pretty huge difference between iOS and Android
![image](https://github.com/firebase/firebase-android-sdk/assets/7146902/55a65d11-61d0-4691-8ddd-eb5be0648ae9)
AppId: hu.mol.move

**Comment by rebeccaoskarsson** [comment]
We also experience the exact same issue. We have done the update including a fix for this bug, but no change. We are considering switching to Snowplow due to this. Any updates on this issue before we start working with a different analytics tool?

**Comment by jmartinMone** [comment]
"Ok Google, can you explain this issue ?"

**Comment by alecoopz** [comment]
Hello!

We are experiencing the same issue after disabling screen view tracking on 2 apps. It seems to happen when users do not close the app and it runs on background. Once in a while the event user_engagement fires, renewing the session. This issue impacts all GA4 Engagement and some Session metrics (only in Android platform).

![image](https://github.com/firebase/firebase-android-sdk/assets/153735979/3614c1d6-7871-479e-924d-5d0a34b36c03)

Enabling automatic screen view tracking is not a solution for us.

**Comment by ilhamiprospect** [comment]
Hi, I have the same issue.
Kindly please consider this a serious problem
@argzdev

Thank you

**Comment by rajeshjadidminc** [comment]
Hi,
I have the same issue with Firebase SDK BOM version: 32.5.0 with Jetpack Compose.

<img width="961" alt="Screenshot 2023-12-15 at 1 04 02 PM" src="https://github.com/firebase/firebase-android-sdk/assets/95206513/8c8fa4c7-d194-43ca-ad14-6f5877a88f5c">

Is there any update on this issue?

Thanks!

**Comment by argzdev** [comment]
Hi folks, sorry for the radio silence here. I've been trying to get in touch with our engineering team and request for updates. So far our engineers are aware of this issue and have been keeping this issue on their priority list. However, they're also occupied with other tasks at hand, so things are taking longer to figure out. Rest assured, we'll share updates once we have progress in our findings.

**Comment by woodii** [comment]
@argzdev any updates on this? I am really curious that such a major issue which makes analysis of your project much harder or even pointless has such a low priority for over 2 years.
Is there anything we can do to help you fix this?

**Comment by alecoopz** [comment]
@argzdev Hello, do you have any update? We are GA 360 users and from what we've analyzed this event is consuming a big percentage of our monthly event contract. By the way, we're considering to change to another tool in the next year because of this experience. There is any way to prioritize this issue?

**Comment by lbarrettanderson** [comment]
> @argzdev Hello, do you have any update? We are GA 360 users and from what we've analyzed this event is consuming a big percentage of our monthly event contract. By the way, we're considering to change to another tool in the next year because of this experience. There is any way to prioritize this issue?

Can you share with me your App ID(s)? We're looking into this again.

**Comment by woodii** [comment]
@lbarrettanderson 1:165995409996:android:334b06cecc30a1ba
Please let me know if we can help in any other way

**Comment by alecoopz** [comment]
> > @argzdev Hello, do you have any update? We are GA 360 users and from what we've analyzed this event is consuming a big percentage of our monthly event contract. By the way, we're considering to change to another tool in the next year because of this experience. There is any way to prioritize this issue?
> 
> Can you share with me your App ID(s)? We're looking into this again.

Of course, here it is:
1:590874711807:android:52ff104a705b789d

**Comment by lbarrettanderson** [comment]
The fix is rolled out. I've checked and I'm seeing significant improvement as of yesterday and today. I expect the improvement will continue and your numbers will look more or less reasonable by early next week.

**Comment by lbarrettanderson** [comment]
> Hi @argzdev Please any updates on this?
> 
> Almost same issue here. But we did NOT [disable](https://firebase.google.com/docs/analytics/screenviews#disable_screenview_tracking) automatic screen_view tracking, but beyond automatic screen_view tracking we are tracking manually screen views.
> 
> <img alt="Snímek obrazovky 2024-01-16 v 15 33 29" width="1165" src="https://private-user-images.githubusercontent.com/117902407/297086800-bbae55ad-932b-46f5-a8d3-2bc36203df01.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU0NDU1NjMsIm5iZiI6MTcxNTQ0NTI2MywicGF0aCI6Ii8xMTc5MDI0MDcvMjk3MDg2ODAwLWJiYWU1NWFkLTkzMmItNDZmNS1hOGQzLTJiYzM2MjAzZGYwMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxMVQxNjM0MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMDA4NTNiM2ZiOGY1NGU2NTE2YmRhMjk4OWRmYjEzYTQ0ODA3MGM5OTIzNDY0MjM3OTNkNTVkODNkNWQxN2YzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.jSvpifIZviHqNAmEbbYkMypiNq94VPBb0CblLR2kZfY">

Since you haven't disabled automatic screen views, your issue might be different. By chance did you either: 1.) Implement consent mode or 2.) start calling the resetAnalyticsData method?

We discovered another engagement-related bug around those two features. If you implemented consent mode, you should see an improvement next week. If you're manually calling resetAnalyticsData, a fix for that issue is coming (timeframe not known, but likely in the next 6 weeks).

**Comment by woodii** [comment]
@lbarrettanderson thank you so much! A first glance at the stats looks promising. 
![image](https://github.com/firebase/firebase-android-sdk/assets/1383203/621c483e-0e6e-45f5-ba24-34b23bad7602)

**Comment by michalnovacek96** [comment]
@lbarrettanderson 
It seems that stats are starting to look ok. But still Android is way much higher than iOS. Avereage time per session on iOS is 46 seconds, on Android 55 minutes. This can't be real.

My setup is:
- I have not **disabled** automatic screen view tracking
- I have setup Google Consent Mode v1. I am populating only analytics_storage.
![Snímek obrazovky 2024-05-21 v 10 34 36](https://github.com/firebase/firebase-android-sdk/assets/117902407/bdaa8224-ddc0-4f5b-acb1-69ab1f85e908)

**Comment by lbarrettanderson** [comment]
> @lbarrettanderson It seems that stats are starting to look ok. But still Android is way much higher than iOS. Avereage time per session on iOS is 46 seconds, on Android 55 minutes. This can't be real.
> 
> My setup is:
> 
> * I have not **disabled** automatic screen view tracking
> * I have setup Google Consent Mode v1. I am populating only analytics_storage.
>   ![Snímek obrazovky 2024-05-21 v 10 34 36](https://private-user-images.githubusercontent.com/117902407/332348056-bdaa8224-ddc0-4f5b-acb1-69ab1f85e908.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMTQ0ODYsIm5iZiI6MTcxNjMxNDE4NiwicGF0aCI6Ii8xMTc5MDI0MDcvMzMyMzQ4MDU2LWJkYWE4MjI0LWRkYzAtNGY1Yi1hY2IxLTY5YWIxZjg1ZTkwOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyMVQxNzU2MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYjgyN2U5ODI3YjYxODhlMmNlZWY4NjllZDNkZjVlNjRkMmRkN2VkMDEzYTM1NGVjYTBmZDY3NGI0MWJkMzZhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.lECTRTTuhyAavH5ZFsiWxtjdZufoQ5ev1Q6BYIhTIvE)

This issue has also been recently fixed, but the recovery time might take longer due to the nature of the bug.

Do you by chance ever call resetAnalyticsData?

**Comment by michalnovacek96** [comment]
@lbarrettanderson No, we don't call resetAnalyticsData

**Comment by lbarrettanderson** [comment]
> @lbarrettanderson No, we don't call resetAnalyticsData

Do you distribute to non-Google Play devices?

**Comment by michalnovacek96** [comment]
@lbarrettanderson No we only distribute Android on Google Play.

**Comment by lbarrettanderson** [comment]
If you're not seeing as much improvement as expected, I recommend updating your SDK to version 22.0.1 or newer. This should help any users who for whatever reason are not getting the updated version of google play services.

**Comment by lbarrettanderson** [comment]
Has anyone who was hoping for more of a reduction in engagement times tried upgrading to firebase-analytics 22.0.1? Did it help?

**Comment by cahyoardhi** [comment]
Any update to this issue?

**Comment by michalnovacek96** [comment]
Hi @lbarrettanderson I can see change, but the reported data is not still trust-worthy. Android reports 5.5 min per session, on the other hand iOS reports less than 1 minute. I can't trust such a data on Android, it seems that issue is still there.

Our firebase-analytics version is 22.0.2

![Snímek obrazovky 2024-09-24 v 13 45 02](https://github.com/user-attachments/assets/e16e9a38-0c2f-424b-b757-33854b3cc99e)

**Comment by tomyeunghb** [comment]
Updated Firebase BOM 33.1.0 around June. But still got high engagement time on the GA4 console (max 28mins, min 7mins).
Agreed that I can see the change, but the value is still doesn't make sense when comparing with iOS

![Screenshot 2024-09-25 at 9 56 36 AM](https://github.com/user-attachments/assets/ad0a4c6d-1c41-405c-bb46-1faa8322c281)

**Comment by lehcar09** [comment]
Hi folks! this issue has been resolved in [Analytics v22.0.1](https://firebase.google.com/support/release-notes/android#analytics_v22-0-1). 

With that, I'll be closing this issue now. Let us know if there’s any misunderstanding so we can re-open the issue. Or, Please feel free to open a new issue if this still occurs. Thanks!

**Comment by tomyeunghb** [comment]
Hi @lehcar09 , pls check [this comment](https://github.com/firebase/firebase-android-sdk/issues/3240#issuecomment-2372731004), already updated to corresponding Firebase BOM but still facing the issue. I don't think the issue has been resolved.
